pinot-core/src/main/java/org/apache/pinot/core/query/reduce/BrokerReduceService.java

```diff
@@ -78,61 +159,69 @@ public BrokerResponseNative reduceOnDataTable(BrokerRequest brokerRequest,
       Map<String, String> metadata = dataTable.getMetadata();
 
       // Reduce on trace info.
-      if (brokerRequest.isEnableTrace()) {
+      if (enableTrace) {
         brokerResponseNative.getTraceInfo()
-            .put(entry.getKey().getHostname(), metadata.get(DataTable.TRACE_INFO_METADATA_KEY));
+            .put(entry.getKey().getHostname(), metadata.get(MetadataKey.TRACE_INFO.getName()));
       }
 
       // Reduce on exceptions.
-      for (String key : metadata.keySet()) {
-        if (key.startsWith(DataTable.EXCEPTION_METADATA_KEY)) {
-          processingExceptions.add(new QueryProcessingException(Integer.parseInt(key.substring(9)), metadata.get(key)));
-        }
+      Map<Integer, String> exceptions = dataTable.getExceptions();
+      for (int key : exceptions.keySet()) {
+        processingExceptions.add(new QueryProcessingException(key, exceptions.get(key)));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.